File: src/main/java/org/jukeboxmc/block/behavior/BlockBed.java
package org.jukeboxmc.block.behavior;

import org.cloudburstmc.nbt.NbtMap;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.BlockType;
import org.jukeboxmc.block.data.BlockColor;
import org.jukeboxmc.block.direction.BlockFace;
import org.jukeboxmc.block.direction.Direction;
import org.jukeboxmc.blockentity.BlockEntity;
import org.jukeboxmc.blockentity.BlockEntityBed;
import org.jukeboxmc.blockentity.BlockEntityType;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.math.Location;
import org.jukeboxmc.math.Vector;
import org.jukeboxmc.player.Player;
import org.jukeboxmc.util.Identifier;
import org.jukeboxmc.world.World;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockBed extends Block implements Waterlogable {

    public BlockBed( Identifier identifier ) {
        super( identifier );
    }

    public BlockBed( Identifier identifier, NbtMap blockStates ) {
        super( identifier, blockStates );
    }

    @Override
    public boolean placeBlock( Player player, World world, Vector blockPosition, Vector placePosition, Vector clickedPosition, Item itemInHand, BlockFace blockFace ) {
        BlockColor blockColor = BlockColor.values()[itemInHand.getMeta()];

        Block blockDirection = this.getSide( player.getDirection() );
        Location directionLocation = blockDirection.getLocation();

        if ( blockDirection.canBeReplaced( this ) && blockDirection.isTransparent() ) {
            if (world.getBlock(placePosition) instanceof BlockWater blockWater && blockWater.getLiquidDepth() == 0 &&
                    world.getBlock(directionLocation) instanceof BlockWater blockWaterDirection && blockWaterDirection.getLiquidDepth() == 0) {
                world.setBlock(placePosition, Block.create(BlockType.WATER), 1, false);
                world.setBlock(directionLocation, Block.create(BlockType.WATER), 1, false);
            }

            this.setDirection( player.getDirection() );

            BlockBed blockBed = Block.create( BlockType.BED );
            blockBed.setLocation( directionLocation );
            blockBed.setDirection( this.getDirection() );
            blockBed.setHeadPiece( true );

            world.setBlock( directionLocation, blockBed, 0 );
            world.setBlock( placePosition, this, 0 );

            BlockEntity.<BlockEntityBed>create( BlockEntityType.BED, blockBed ).setColor( blockColor ).spawn();
            BlockEntity.<BlockEntityBed>create( BlockEntityType.BED, this ).setColor( blockColor ).spawn();
            return true;
        }
        return false;
    }

    @Override
    public void onBlockBreak( Vector breakPosition ) {
        Direction direction = this.getDirection();
        if ( this.isHeadPiece() ) {
            direction = direction.opposite();
        }
        Block otherBlock = this.getSide( direction );
        if ( otherBlock instanceof BlockBed blockBed ) {
            if ( blockBed.isHeadPiece() != this.isHeadPiece() ) {
                this.location.getWorld().setBlock( otherBlock.getLocation(), Block.create( BlockType.AIR ), 0 );
            }
        }
        this.location.getWorld().setBlock( breakPosition, Block.create( BlockType.AIR ), 0 );
    }

    @Override
    public BlockEntityBed getBlockEntity() {
        return (BlockEntityBed) this.location.getWorld().getBlockEntity( this.location, this.location.getDimension() );
    }

    @Override
    public int getWaterLoggingLevel() {
        return 1;
    }

    public void setHeadPiece(boolean value ) {
        this.setState( "head_piece_bit", value ? (byte) 1 : (byte) 0 );
    }

    public boolean isHeadPiece() {
        return this.stateExists( "head_piece_bit" ) && this.getByteState( "head_piece_bit" ) == 1;
    }

    public void setOccupied( boolean value ) {
        this.setState( "occupied_bit", value ? (byte) 1 : (byte) 0 );
    }

    public boolean isOccupied() {
        return this.stateExists( "occupied_bit" ) && this.getByteState( "occupied_bit" ) == 1;
    }

    public void setDirection( Direction direction ) {
        switch ( direction ) {
            case SOUTH -> this.setState( "direction", 0 );
            case WEST -> this.setState( "direction", 1 );
            case NORTH -> this.setState( "direction", 2 );
            case EAST -> this.setState( "direction", 3 );
        }
    }

    public Direction getDirection() {
        int value = this.stateExists( "direction" ) ? this.getIntState( "direction" ) : 0;
        return switch ( value ) {
            case 0 -> Direction.SOUTH;
            case 1 -> Direction.WEST;
            case 2 -> Direction.NORTH;
            default -> Direction.EAST;
        };
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockBed.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.Bed
import org.jukeboxmc.api.block.Block
import org.jukeboxmc.api.block.data.BlockFace
import org.jukeboxmc.api.block.data.Color
import org.jukeboxmc.api.block.data.Direction
import org.jukeboxmc.api.blockentity.BlockEntity
import org.jukeboxmc.api.blockentity.BlockEntityBed
import org.jukeboxmc.api.blockentity.BlockEntityType
import org.jukeboxmc.api.math.Vector
import org.jukeboxmc.server.block.JukeboxBlock
import org.jukeboxmc.server.extensions.toByte
import org.jukeboxmc.server.extensions.toJukeboxBlock
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer
import org.jukeboxmc.server.world.JukeboxWorld

class BlockBed(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates),
    Bed {

    override fun placeBlock(
        player: JukeboxPlayer,
        world: JukeboxWorld,
        blockPosition: Vector,
        placePosition: Vector,
        clickedPosition: Vector,
        itemInHand: JukeboxItem,
        blockFace: BlockFace
    ): Boolean {
        val blockColor = Color.entries[itemInHand.getMeta()]
        val direction = player.getDirection()

        val blockDirection = this.getRelative(direction.toBlockFace()).toJukeboxBlock()
        val directionLocation = blockDirection.getLocation()

        if (blockDirection.canBeReplaced(this) && blockDirection.isTransparent() && !this.getRelative(BlockFace.DOWN)
                .isTransparent() && !directionLocation.getBlock().getRelative(BlockFace.DOWN).isTransparent()
        ) {
            this.setDirection(direction)

            val blockBed: BlockBed = this.clone() as BlockBed
            blockBed.setHeadPiece(true)

            world.setBlock(directionLocation, blockBed)
            world.setBlock(placePosition, this)

            BlockEntity.create<BlockEntityBed>(BlockEntityType.BED, blockBed)?.let {
                it.setColor(blockColor)
                it.spawn()
            }

            BlockEntity.create<BlockEntityBed>(BlockEntityType.BED, this)?.let {
                it.setColor(blockColor)
                it.spawn()
            }
            return true
        }
        return false
    }

    override fun onBlockBreak(breakLocation: Vector) {
        var direction = getDirection()
        if (this.isHeadPiece()) {
            direction = direction.opposite()
        }
        val otherBlock: Block = this.getRelative(direction.toBlockFace())
        if (otherBlock is BlockBed) {
            if (otherBlock.isHeadPiece() != this.isHeadPiece()) {
                this.getWorld().setBlock(otherBlock.getLocation(), AIR)
            }
        }
        this.getWorld().setBlock(breakLocation, AIR)
    }

    override fun getDirection(): Direction {
        return when (this.getIntState("direction")) {
            0 -> Direction.SOUTH
            1 -> Direction.WEST
            2 -> Direction.NORTH
            else -> Direction.EAST
        }
    }

    override fun setDirection(value: Direction): BlockBed {
        return when (value) {
            Direction.SOUTH -> this.setState("direction", 0)
            Direction.WEST -> this.setState("direction", 1)
            Direction.NORTH -> this.setState("direction", 2)
            else -> this.setState("direction", 3)
        }
    }

    override fun isHeadPiece(): Boolean {
        return this.getBooleanState("head_piece_bit")
    }

    override fun setHeadPiece(value: Boolean): BlockBed {
        return this.setState("head_piece_bit", value.toByte())
    }

    override fun isOccupied(): Boolean {
        return this.getBooleanState("occupied_bit")
    }

    override fun setOccupied(value: Boolean): BlockBed {
        return this.setState("occupied_bit", value.toByte())
    }
}